zeroize/src/lib.rs

```diff
@@ -325,6 +325,41 @@ where
     /// previous reallocations did not leave values on the heap.
     fn zeroize(&mut self) {
         self.iter_mut().zeroize();
+
+        // Zero the capacity of the `Vec` that is not initialized.
+        {
+            // Safety:
+            //
+            // This is safe, because `Vec` never allocates more than `isize::MAX` bytes.
+            // This exact use case is even mentioned in the documentation of `pointer::add`.
+            let extra_capacity_start = unsafe { self.as_mut_ptr().add(self.len()) as *mut u8 };
+            let extra_capacity_len = self.capacity().saturating_sub(self.len());
+
+            // Safety:
+            //
+            // There are three invariants that need to be fulfilled:
+            // - The data must be valid for the entire length and must be properly aligned.
+            //       This is true, because the `Vec` previously reserved `self.capacity() *
+            //       mem::size_of::<Z>()` bytes and the length used here is always less than or
+            //       equal to that.
+            //       The alignment is always proper, because `u8` has an alignment of `1`.
+            // - The memory of the slice must not be referenced through any other pointer during
+            //   its lifetime.
+            //       This is true, since we have a mutable reference to the `Vec`, which ensures
+            //       that we have exclusive access to it and we do not use it in any other way,
+            //       while `extra_capacity` is in scope.
+            // - The total size of the slice must be no larger than `isize::MAX`.
+            //       This is true, because `Vec` never allocates more than `isize::MAX` bytes.
+            let extra_capacity = unsafe {
+                core::slice::from_raw_parts_mut(
```

Comment:
This appears to create a slice of uninitialized memory, to give it to the `Zeroize` impl on `[Z]`, which iterates over it and performs `ptr::write_volatile`.

It might make sense instead to perform `ptr::write_volatile` here directly, followed by `atomic_fence()`, to avoid ever constructing an uninitialized slice.

Reply:
Alternatively this was recently stabilized: https://github.com/rust-lang/rust/pull/68234

Reply:
I don't quite see how constructing a `*mut [T]` helps. It's main intent is for data structures storing a `NonNull<T>`  but here it would still require somehow manually iterating over the individual element pointers to do a volatile write. In particular, the implementation of `Z::zeroize(&mut self)` can still not be called without UB.